### PR TITLE
Track executing requests

### LINKED ISSUE #118746: Debug endpoint to dump currently executing requests
### What would you like to be added?

The currently available debug endpoints for APF are for dump_priority_levels, dump_queues, and dump_requests. These commands provide useful information primarily about the state of APF queues, including the the FlowDistinguisher, ArriveTime, UserName, Verb, and APIPath of requests waiting in queues. To help debug APF-related issues, it would be valuable to provide similar information about requests which have been dispatched by APF and are currently occupying 1 or more seats. The existing commands for dump_priority_levels and dump_queues give a count for the number of executing requests per PriorityLevelConfiguration or queue but there is not enough information available to identify which requests those are without also using information available from audit or apiserver logs.

### Why is this needed?

This new debug command, which could be called something like dump_executing_requests, could assist with live-debugging of APF issues. Situations where this would be useful:
1. General APF debugging to see which problematic requests are using a majority of seats per PriorityLevelConfiguration. This would aid in live-debugging in situations where the given request type is not included in the cluster's audit policy or if verbose apiserver logging is not enabled. Situations where a relatively low number of requests use a large number of seats could be debugged quicker (for example if listing 100k+ event objects is taking around 60 seconds).
2. APF issues with stuck requests. For example, this previous issue with broken CRD could've been quickly reproduced if we knew which watch request was stuck executing, https://github.com/kubernetes/kubernetes/issues/109603. A related issue that is still possible is issuing a watch request against a resource with an uninitialized watch-cache and this request will occupy an inflight request until the client drops the request or the watch-cache initializes. The audit event for the watch request will not be available until the request completes. To see which watch request is stuck, we would have to check apiserver logs at verbosity 3+ for requests which were dispatched but never completed.

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR extends API Priority and Fairness to track the individual executing requests and optionally report them through a debug URL.

#### Which issue(s) this PR fixes:

Fixes #118746

#### Special notes for your reviewer:
This PR is not done yet, there is an interface design question in #118746 to be resolved first.

This PR builds on #118955 and will be rebased after that one merges.

#### Does this PR introduce a user-facing change?

```release-note
The apiserver debug endpoint `/debug/api_priority_and_fairness/dump_requests` has been extended to dump executing requests as well as queued ones.  A column for StartTime has been added to the returned table, with the queued requests having a StartTime of "0001-01-01T00:00:00Z".  The executing requests have a RequestIndexInQueue of -1, and the QueueIndex is also -1 for priority levels without queues.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
/cc @wojtek-t 
@tkashem 
/uncc @yue9944882

**Comment by MikeSpreitzer** [comment]
@natherz97

**Comment by natherz97** [comment]
Thank you for taking this up! When possible, can you share example output from the new debug command?

**Comment by MikeSpreitzer** [comment]
@natherz97 : I realized that the design is not obvious (to me, anyway) and posted about that in the issue.  I am waiting for feedback on https://github.com/kubernetes/kubernetes/issues/118746#issuecomment-1615405895

**Comment by MikeSpreitzer** [comment]
@natherz97: See the Release Note in the opening comment here.  The added column for StartTime is inserted right after the AdditionalLatency column.

**Comment by MikeSpreitzer** [comment]
The tracking of executing requests in QueueSet is wrong --- the requests are tracked in each queue, but not every QueueSet has queues.
/hold

**Comment by MikeSpreitzer** [comment]
Bug fixed.
/unhold

**Comment by MikeSpreitzer** [comment]
@natherz97 : here is an example.

```console
root@ubu2204a:/var/run/kubernetes# kubectl get --raw /debug/api_priority_and_fairness/dump_requests
PriorityLevelName, FlowSchemaName, QueueIndex, RequestIndexInQueue, FlowDistingsher, ArriveTime,                     InitialSeats, FinalSeats, AdditionalLatency, StartTime
exempt,            exempt,         -1,         -1,                  ,                2023-07-15T04:51:25.596404345Z, 1,            0,          0s,                2023-07-15T04:51:25.596404345Z
```

**Comment by natherz97** [comment]
> @natherz97 : here is an example.
> 
> ```
> root@ubu2204a:/var/run/kubernetes# kubectl get --raw /debug/api_priority_and_fairness/dump_requests
> PriorityLevelName, FlowSchemaName, QueueIndex, RequestIndexInQueue, FlowDistingsher, ArriveTime,                     InitialSeats, FinalSeats, AdditionalLatency, StartTime
> exempt,            exempt,         -1,         -1,                  ,                2023-07-15T04:51:25.596404345Z, 1,            0,          0s,                2023-07-15T04:51:25.596404345Z
> ```
@MikeSpreitzer thank you! It also looks like dump_requests supports the includeRequestDetails=1 parameter to provide information like the username, verb, and path?

**Comment by MikeSpreitzer** [comment]
Yes.  I just gave an example of the simpler usage.  The extended result's columns are the same as before.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go [hunk 1]
```diff
@@ -175,9 +162,54 @@ func (q *queue) dumpLocked(includeDetails bool) debug.QueueDump {
...
+
+func MapSetToSlice[EltInput comparable, EltOutput any](set sets.Set[EltInput], mapFn func(EltInput) EltOutput) []EltOutput {
+	ans := make([]EltOutput, 0, set.Len())
+	for elt := range set {
+		ans = append(ans, mapFn(elt))
+	}
+	return ans
+}
+
+func OrFuncOverSlice[Elt any](f func(Elt) bool) func([]Elt) bool {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:206]: Given it's a function that is used only in the test, can we move it to that test file? I doubt we will ever remember about this function exists here...

**Review by wojtek-t — COMMENTED** [review]
I have two comments, but overall this looks reasonable to me.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:206]: This is a form of map-reduce, I am sure I will remember that map-reduce exists and that I am tired of writing it over and over.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:206]: Perhaps it should be

```go
func SliceMapReduce[Elt, Result any](mapFn func(Elt) Result, reduceFn func(Result,Result) Result) func([]Elt) Result {...{
```

**Comment by MikeSpreitzer** [comment]
The force-push to a8a2fb317c8 removes the currying function and restates the two map-reduce functions in the general form that hopefully makes the hope for more future use clear.

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**